### ISSUE TITLE
[Βήμα 19] Υλοποίηση NN στη μορφή του sklearn

### ISSUE THREAD

**Issue by foivospar** [slp-ntua/patrec-labs#21]
### Καλησπέρα, παραθέτω συγκεντρωμένες κάποιες απορίες μου από τα docstrings του Βήματος 19:

**1.** Στην init του PytorchNNModel αναφέρονται τα ορίσματα ως (*args, **kwargs). Αυτό σημαίνει :
      (α) ότι το αφήνουμε έτσι, και υλοποιούμε κατάλληλα την init ώστε να "τρέχει" για οποιαδήποτε μορφής ορίσματα (που 
          μπορεί να σημαίνουν layers, αριθμός νευρώνων κλπ) του δώσει κάποιος;
                 ή
      (β) μπορούμε να τα αντικαταστήσουμε με δικά μας ορίσματα, εξηγώντας με σχόλια τι είναι το καθένα;
Ρωτάω διότι το (β) θα διευκόλυνε εμάς ως προς το να αλλάζουμε απλώς τις τιμές των ορισμάτων (και όχι την υλοποίηση της init) όταν θέλουμε να δοκιμάσουμε διαφορετικές αρχιτεκτονικές. Όμως, το (α) θα ήταν αναγκαίο εάν είναι να τρέξετε test cases και για το Βήμα 19.

**2.** Στην fit(X,y) ζητείται να δημιουργήσουμε dataloaders για train και dev set. Φαντάζομαι πως το dev set χρειάζεται για το tuning των υπερπαραμέτρων. Όμως, εφόσον η fit πρέπει απλά να κάνει "return self", πως μπορούμε ταυτόχρονα και να εκπαιδεύσουμε το νευρωνικό, και να υπολογίσουμε κάποια μετρική πάνω στο dev set. Δεν καταλαβαίνω, δηλαδή, πως μας βοηθάει το να γίνεται ο χωρισμός σε train/dev εντός της fit.

**3.** Στην predict(X) ζητείται να δημιουργηθεί αρχικά ένας dataloader για το test set, και μετά να γίνουν οι προβλέψεις. Απ'όσο έχω καταλάβει, οι Dataloader χρησιμεύουν για το shuffling και τον χωρισμό σε mini-batches. Έχει όντως νόημα να γίνει αυτό εδώ που μας ενδιαφέρουν απλά οι προβλέψεις σε όλο το X (και φυσικά χωρίς shuffling); Γιατί να μην περνάμε απλά στο μοντέλο κατευθείαν όλο το X (αφού μετατραπεί σε Tensor) χωρίς κάποιον test loader; 

Ευχαριστώ και συγνώμη για το μακροσκελές του κειμένου.

**Comment by efthymisgeo** [comment]
1. Περιμένουμε το interface ναι είναι το ίδιο με τις κλάσεις που σας δείξαμε στο tutorial. Aπό εκεί και πέρα κάντε όπως σας βολεύει.
2. To dev χρησιμεύει και για early stopping οπότε βλέπεις ταυτόχρονα και to dev/validation loss και τερματίζεις την εκπαίδευση ώστε να μην κάνεις ακραίο overfit.
3. Αν εχεις πολυ μεγαλο test set δεν μπορεις να το φορτώσεις όλο στη μνήμη οπότε έχει νόημα. To shuffling δεν το συνιστω για το test

**Comment by konskoko** [comment]
Πάντως ως προς το 1 λέει κι [εδώ](https://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html) ότι "All estimators should specify all the parameters that can be set at the class level in their \_\_init\_\_ as explicit keyword arguments (no *args or **kwargs)."

Οπότε νομίζω θα πρέπει να το αλλάξουμε αυθαίρετα ή αν θέλετε κάποια συγκεκριμένα ορίσματα εσείς να μας τα προσδιορίσετε.
